src/Interpreters/HashJoin/AddedColumns.cpp

```diff
@@ -17,19 +17,34 @@ JoinOnKeyColumns::JoinOnKeyColumns(
 {
 }
 
-size_t LazyOutput::buildOutput(size_t size_to_reserve,
+size_t LazyOutput::buildOutput(
+    size_t size_to_reserve,
+    const Block & left_block,
+    const IColumn::Offsets & left_offsets,
     MutableColumns & columns,
     const UInt64 * row_refs_begin,
     const UInt64 * row_refs_end,
     size_t rows_offset,
-    size_t rows_limit) const
+    size_t rows_limit,
+    size_t bytes_limit) const
 {
     if (!output_by_row_list)
         buildOutputFromBlocks<false>(size_to_reserve, columns, row_refs_begin, row_refs_end);
     else
     {
         if (rows_limit)
-            return buildOutputFromBlocksLimitAndOffset(columns, row_refs_begin, row_refs_end, rows_offset, rows_limit);
+        {
+            PaddedPODArray<UInt64> left_sizes;
+            if (bytes_limit)
+            {
+                for (const auto & col : left_block)
+                    col.column->collectSerializedValueSizes(left_sizes, nullptr);
```

Comment:
We do this for the full column for every batch iteration (same left_block with different offset/limit). Let's pre-calculate it once?